content/5_Plat/Centroid.mdx

```diff
@@ -302,88 +313,154 @@ private static class Centroid {
-	for (auto x : a[n])
-		if (!r[x] && x != p && s[x] * 2 > ms) return find(x, ms, n);
-	return n;
+
+/**
+ * Calculate the distance between current `node` and the `centroid` it belongs
+ * to. The distances between a node and all its centroid ancestors are stored
+ * in the vector `ancestors`.
+ * @param cur_dist the distance between `node` and `centroid`
+ */
+void get_dists(int node, int centroid, int parent = -1, int cur_dist = 1) {
+	for (int child : adj[node]) {
+		if (child == parent || is_removed[child]) { continue; }
+		++cur_dist;
+		get_dists(child, centroid, node, cur_dist);
+		--cur_dist;
+	}
+	ancestors[node].push_back({centroid, cur_dist});
 }
-void dfs2(int n, int p = 0) {
-	for (auto x : a[n])
-		if (!r[x] && x != p) ++d, dfs2(x, n), --d;
-	g[n].push_back({t, b, d});
+
+void build_centroid_decomp(int node = 0) {
+	int centroid = get_centroid(node, get_subtree_size(node));
+
+	/*
+	 * For all nodes in the subtree rooted at `centroid`, calculate their
+	 * distances to centroid
+	 */
+	for (int child : adj[centroid]) {
+		if (is_removed[child]) { continue; }
+		get_dists(child, centroid, centroid);
+	}
+
+	is_removed[centroid] = true;
+	for (int child : adj[centroid]) {
+		if (is_removed[child]) { continue; }
+		// build the centroid decomposition for all child components
+		build_centroid_decomp(child);
+	}
 }
-void centroid(int n = 1) {
-	n = find(n, dfs(n));
-	v[n].reserve(a[n].size());
-	for (auto x : a[n])
-		if (!r[x]) {
-			t = n, b = v[n].size(), d = 1;
-			dfs2(x, n);
-			v[n].push_back(INF);
-		}
-	r[n] = 1;
-	for (auto x : a[n])
-		if (!r[x]) centroid(x);
+
+/**
+ * Paint `node` red by updating all of its ancestors' minimal distances
+ * to a red node
+ */
+void paint(int node) {
+	for (auto &[ancestor, dist] : ancestors[node]) {
+		min_dist[ancestor] = min(min_dist[ancestor], dist);
+	}
+	min_dist[node] = 0;
+}
+
+/** Print the minimal distance between `node` to a red node */
+void query(int node) {
+	int ans = min_dist[node];
+	for (auto &[ancestor, dist] : ancestors[node]) {
+		if (!dist) { continue; }
+		/*
+		 * The distance between `node` and a red painted node is the sum of
+		 * the distance from `node` to one of its ancestors (`dist`) and the
+		 * distance from this ancestor to the nearest red node
+		 * (`min_dist[ancestor]`).
+		 */
+		ans = min(ans, dist + min_dist[ancestor]);
+	}
+
+	cout << ans << "\n";
 }
 
 int main() {
-	scanf("%d%d", &N, &M);
-	for (int i = 0, u, v; i + 1 < N; ++i)
-		scanf("%d%d", &u, &v), a[u].push_back(v), a[v].push_back(u);
-	centroid();
-	memset(m, 0x3f, sizeof m);
-	for (int i = 0, t = 1, n = 1; i <= M; ++i) {
+	int N, M;
+	cin >> N >> M;
+
+	adj.assign(N, vector<int>());
+	for (int i = 0, a, b; i < N - 1; ++i) {
+		cin >> a >> b;
+		a--, b--;
+		adj[a].push_back(b);
+		adj[b].push_back(a);
+	}
+
+	subtree_size.assign(N, 0);
+	ancestors.assign(N, vector<pair<int, int>>());
+	is_removed.assign(N, false);
+	build_centroid_decomp();
+
+	min_dist.assign(N, INF);
```

Comment:
wait i forgor why you use inf & not int32_max

Reply:
cause it will overflow when trying to add up two distances